utils/sites.js

```diff
@@ -282,37 +254,99 @@ const updateParticipantData = async (req, res, authObj) => {
             await siteNotificationsHandler(docData['Connect_ID'], '987563196', docData['827220437'], obj);
         }
 
+        // Flatten dataObj and docData for comparison & validation.
+        const { flattened: flatDataObj, arrayPaths: flatArrayPaths } = flattenObject(dataObj);
+        const { flattened: flatDocData } = flattenObject(docData);
+
+        // Delete primary identifiers from dataObj if they exist. There is no reason to update these through this API.
+        const primaryIdentifiers = ['token', 'pin', 'Connect_ID', 'state.uid'];
+        for (const identifier of primaryIdentifiers) {
+            delete flatDataObj[identifier];
+        }
+
+        // Validate incoming data. flatDocData is only used to check the 'mustExist' property in some rules.
         if(!authObj) {
-            const errors = qc(updatedData, flattened.docData, rules);
+            const errors = flatValidationHandler(flatDataObj, flatDocData, rules, validateUpdateData);
             if(errors.length !== 0) {
                 error = true;
                 responseArray.push({'Invalid Request': {'Token': participantToken, 'Errors': errors}});
                 continue;
             }
         }
 
+        // Check initializeTimestamps and init on match. Currently, only one key exists in initializeTimestamps.
         const { initializeTimestamps } = require('./shared')
+        const keysForTimestampGeneration = Object.keys(initializeTimestamps);
+        for (const key of keysForTimestampGeneration) {
+            if (initializeTimestamps[key] && flatDataObj[key] != null) {
+                if (initializeTimestamps[key].value && initializeTimestamps[key].value === flatDataObj[key]) {
+                    Object.assign(flatDataObj, initializeTimestamps[key].initialize);
+                }
+            }
+        }
 
-        for(let key in updatedData) {
-            if(initializeTimestamps[key]) {
-                if(initializeTimestamps[key].value && initializeTimestamps[key].value === updatedData[key]) {
-                    updatedData = {...updatedData, ...initializeTimestamps[key].initialize}
+        // Handle the array paths: Firestore writes are nuanced for arrays. We either need to manage the array or use FieldValue.arrayUnion & FieldValue.arrayRemove.
```

Comment:
would like to hop on a call to discuss this whenever you're free

Reply:
Thanks for working through this on the phone.